Mathlib/Computability/MyhillNerode.lean

```diff
@@ -0,0 +1,98 @@
+/-
+Copyright (c) 2024 Google. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Chris Wong
+-/
+import Mathlib.Computability.DFA
+import Mathlib.Data.Set.Finite.Basic
+
+/-!
+# Myhill–Nerode theorem
+
+This file proves the Myhill–Nerode theorem using left quotients.
+
+Given a language `L` and a word `x`, the *left quotient* of `L` by `x` is the set of suffixes `y`
+such that `x ++ y` is in `L`. The *Myhill–Nerode theorem* shows that each left quotient, in fact,
+corresponds to the state of an automaton that matches `L`, and that `L` is regular if and only if
+there are finitely many such states.
+
+## References
+
+* <https://en.wikipedia.org/wiki/Syntactic_monoid#Myhill%E2%80%93Nerode_theorem>
+-/
+
+universe u v
+variable {α : Type u} {σ : Type v} (L : Language α)
```

Comment:
```suggestion
variable {α : Type u} {σ : Type v} {L : Language α}
```
and can you make it explicit where necessary using `variable (L) in`?

Reply:
Sounds good.

Do we have a convention for when to make the `L` explicit?

So far I've only made it explicit for definitions. I'm ok with that, but I wonder if you have an opinion too.

Reply:
Guideline is:
* implicit for iffs
* explicit for definitions
* explicit for rewriting lemmas other than iffs
* implicit for non-rewriting lemmas 

with some leeway in each case

Reply:
Ok, I made it explicit in all the equality lemmas where the `L` isn't already fixed by something else. PTAL!